src/chainstate/stacks/boot/pox.clar

```diff
@@ -0,0 +1,754 @@
+        (asserts! (is-none (get-stacker-info delegate))
+            (err ERR-STACKING-ALREADY-DELEGATED))
+
+        ;; the delegate's first reward cycle must not have started yet
+        (asserts! (< (current-pox-reward-cycle) (get first-reward-cycle delegate-control-info))
+            (err ERR-STACKING-DELEGATE-EXPIRED))
+
+        ;; the Stacker must have sufficient unlocked funds
+        (asserts! (>= (stx-get-balance tx-sender) amount-ustx)
+            (err ERR-STACKING-INSUFFICIENT-FUNDS))
+
+        ;; encode the delegate/client relationship
+        (map-set delegates
+            { stacker: tx-sender }
+            {
+                delegate: delegate,
+                amount-ustx: amount-ustx
+            }
+        )
+
+        ;; update how much uSTX the delegate now controls in total
+        (map-set delegate-control
+            { delegate: delegate }
+            {
+                total-ustx: (+ amount-ustx (get total-ustx delegate-control-info)),
+                pox-addr: (get pox-addr delegate-control-info),
+                burn-block-height-start: (get burn-block-height-start delegate-control-info),
+                first-reward-cycle: (get first-reward-cycle delegate-control-info),
+                tenure-cycles: (get tenure-cycles delegate-control-info)
+            }
+        )
+
+        ;; add a stacker record for this principal
+        (map-set stacking-state
+            { stacker: tx-sender }
+            {
+                amount-ustx: amount-ustx,
+                pox-addr: (get pox-addr delegate-control-info),
+                first-reward-cycle: first-reward-cycle,
+                lock-period: (get tenure-cycles delegate-control-info),
+                delegate: (some delegate)
+            }
+        )
+
+        ;; we're done! let the delgate do its thing.
+        ;; Give back the information the node needs to actually carry out the lock.
+        (ok { stacker: tx-sender, lock-amount: amount-ustx, unlock-burn-height: (reward-cycle-to-burn-height (+ first-reward-cycle (get tenure-cycles delegate-control-info))) })
+    ))
+)
+
+;; Lock up some uSTX for stacking!  Note that the given amount here is in micro-STX (uSTX).
+;; The STX will be locked for the given number of reward cycles (lock-period).
+;; This is the self-service interface.  tx-sender will be the Stacker.
+;;
+;; * The given stacker cannot currently be stacking.
+;; * You will need the minimum uSTX threshold.  This will be determined by (get-stacking-minimum)
+;; at the time this method is called.
+;; * You may need to increase the amount of uSTX locked up later, since the minimum uSTX threshold
+;; may increase between reward cycles.
+;;
+;; The tokens will unlock and be returned to the Stacker (tx-sender) automatically.
+(define-public (stack-stx (amount-ustx uint)
+                          (pox-addr (tuple (version uint) (hashbytes (buff 20))))
+                          (lock-period uint))
```

Comment:
Can we split this method in 2, so that users can check their eligibility through a `read-only` call?

Reply:
when a user is stacking and invoking this method, should he specify the starting reward-cycle?

Reply:
> Can we split this method in 2, so that users can check their eligibility through a read-only call?

Sure, I can do this.

> when users are stacking and invoking this method, should they be specifying the starting reward-cycle themselves?

￼I don't think so.  The users' tokens must be locked up immediately after this call, and the tokens can only be used in at most 12 reward cycles before they must be re-locked.  Allowing the user to specify the reward cycle as something _other than_ the very next reward cycle would require coming up with a way for the user's tokens to be locked, but somehow not involved in PoX until the desired reward cycle (and, a misplaced argument for the reward cycle length could render the STX locked for an arbitrarily long amount of time, waiting for a reward cycle that is far in the future).

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/1850/commits/9d35d058d3df40dc2a045144c19fe389ddfe7e62